benchmarks/llama.py

```diff
@@ -0,0 +1,236 @@
+          "vocab_size": -1
+      },
+      '13b.json': {
+          "dim": 5120,
+          "multiple_of": 256,
+          "n_heads": 40,
+          "n_layers": 40,
+          "norm_eps": 1e-05,
+          "vocab_size": -1
+      },
+      '70b.json': {
+          "dim": 8192,
+          "multiple_of": 4096,
+          "ffn_dim_multiplier": 1.3,
+          "n_heads": 64,
+          "n_kv_heads": 8,
+          "n_layers": 80,
+          "norm_eps": 1e-05,
+          "vocab_size": -1
+      }
+  }
+  for filename, config in model_configs.items():
+    filepath = os.path.join(llama_dir, filename)
+    with open(filepath, 'w') as f:
+      json.dump(config, f)
+  return llama_dir
+
+
+def parse_log_file(log_file: str):
+  latencies = []
+  with open(log_file, 'r') as f:
+    for line in f:
+      if ('Totally decoded ' not in line or 'tokens in' not in line or
+          ' seconds' not in line):
+        continue
+      parts = line.strip().split()
+      tokens = float(parts[2])
+      seconds = float(parts[5])
+      latency_per_token = seconds / tokens
+      latencies.append(latency_per_token)
+  return latencies
+
+
+def benchmark_has_already_run(results_file: str, model_name: str, xla: str,
+                              dynamo: str, batch_size: int):
+  with open(results_file, 'r') as f:
+    for line in f:
+      # Grep for relevant lines to avoid parsing the entire JSONL file.
+      if f'"model_name": "{model_name}"' not in line:
+        continue
+      r = json.loads(line.rstrip('\n|\r'))
+      # yapf: disable
+      if all(
+          r.get(k1, {}).get(k2) == v
+          for (k1, k2, v) in [
+              ('experiment', 'accelerator', 'cuda'),
+              ('experiment', 'batch_size', batch_size),
+              ('experiment', 'dynamo', dynamo),
+              ('experiment', 'test', 'eval'),
+              ('experiment', 'xla', xla),
+              ('experiment', 'xla_flags', None),
+              ('model', 'model_name', model_name),
+          ]):
+        return True
+      # yapf: enable
+  return False
+
+
+def run_benchmarks(args, llama_dir: str, results_dir: str,
+                   accelerator_model: str, timestamp: float):
+  os.chdir(llama_dir)
+  for size in ['7b', '13b', '70b']:
+    params_json = 'params.json'
+    if os.path.exists(params_json):
+      os.remove(params_json)
+    os.symlink(f'{size}.json', params_json)
+    model_name = f"llama2.{size}"
+    for dynamo in [None, 'inductor', 'openxla', 'openxla_eval']:
+      backend = dynamo if dynamo else 'lazytensor'
+      xla = None if dynamo == 'inductor' else 'PJRT'
+      summary = f"{model_name} eval {backend} batch {args.batch_size}"
+
+      results_file = os.path.join(results_dir, 'results.jsonl')
+      if benchmark_has_already_run(results_file, model_name, xla, dynamo,
+                                   args.batch_size):
+        print(f"SKIP already completed benchmark -- {summary}")
+        continue
+
+      print(f"RUN {summary}")
```

Comment:
Should we use a logger?

Reply:
I've added a bunch of debug logging printouts and also added the --log-level flag, just like in experiment_runner.py.